File: snippets/screens/flutter/live_stream_screen/1/live_stream_screen.dart
import 'package:flutter/material.dart';

class LiveStreamScreen extends StatefulWidget {
  final String? streamerUsername;
  final bool isStreaming;

  const LiveStreamScreen({
    Key? key,
    this.streamerUsername,
    this.isStreaming = false,
  }) : super(key: key);

  @override
  _LiveStreamScreenState createState() => _LiveStreamScreenState();
}

class _LiveStreamScreenState extends State<LiveStreamScreen>
    with SingleTickerProviderStateMixin {
  final TextEditingController _messageController = TextEditingController();
  final ScrollController _chatController = ScrollController();
  late AnimationController _liveIndicatorController;

  int viewerCount = 1247;
  int likeCount = 3456;
  bool isLiked = false;
  bool isMuted = false;
  bool isStreaming = false;
  bool showChat = true;

  List<LiveMessage> messages = [
    LiveMessage(
      id: '1',
      username: 'viewer1',
      message: 'Great stream! 🔥',
      timestamp: DateTime.now().subtract(Duration(minutes: 2)),
      isHighlighted: false,
    ),
    LiveMessage(
      id: '2',
      username: 'viewer2',
      message: 'Hello everyone! 👋',
      timestamp: DateTime.now().subtract(Duration(minutes: 1)),
      isHighlighted: false,
    ),
    LiveMessage(
      id: '3',
      username: 'moderator',
      message: 'Welcome to the stream!',
      timestamp: DateTime.now().subtract(Duration(seconds: 30)),
      isHighlighted: true,
    ),
  ];

  @override
  void initState() {
    super.initState();
    isStreaming = widget.isStreaming;

    _liveIndicatorController = AnimationController(
      duration: Duration(milliseconds: 1000),
      vsync: this,
    );

    if (isStreaming) {
      _liveIndicatorController.repeat();
    }

    // Simulate new messages
    _simulateMessages();

    // Auto-scroll chat to bottom
    WidgetsBinding.instance.addPostFrameCallback((_) {
      _scrollToBottom();
    });
  }

  void _simulateMessages() {
    if (!isStreaming) return;

    Future.delayed(Duration(seconds: 5), () {
      if (mounted && isStreaming) {
        setState(() {
          messages.add(LiveMessage(
            id: DateTime.now().millisecondsSinceEpoch.toString(),
            username: 'random_viewer',
            message: 'This is amazing! 😍',
            timestamp: DateTime.now(),
            isHighlighted: false,
          ));
          viewerCount += (DateTime.now().millisecond % 5);
        });
        _scrollToBottom();
        _simulateMessages();
      }
    });
  }

  @override
  Widget build(BuildContext context) {
    return Scaffold(
      backgroundColor: Colors.black,
      body: Stack(
        children: [
          // Video stream area
          Container(
            width: double.infinity,
            height: double.infinity,
            child: _buildVideoStream(),
          ),
          // Top overlay with streamer info
          SafeArea(
            child: _buildTopOverlay(),
          ),
          // Chat overlay
          if (showChat) _buildChatOverlay(),
          // Bottom controls
          Positioned(
            bottom: 0,
            left: 0,
            right: 0,
            child: SafeArea(
              child: _buildBottomControls(),
            ),
          ),
        ],
      ),
    );
  }

  Widget _buildVideoStream() {
    if (isStreaming) {
      return Container(
        decoration: BoxDecoration(
          gradient: LinearGradient(
            begin: Alignment.topCenter,
            end: Alignment.bottomCenter,
            colors: [
              Colors.purple.withOpacity(0.3),
              Colors.black,
              Colors.blue.withOpacity(0.3),
            ],
          ),
        ),
        child: Center(
          child: Column(
            mainAxisAlignment: MainAxisAlignment.center,
            children: [
              Icon(
                Icons.videocam,
                size: 120,
                color: Colors.white.withOpacity(0.5),
              ),
              SizedBox(height: 20),
              Text(
                'Live Stream',
                style: TextStyle(
                  color: Colors.white,
                  fontSize: 24,
                  fontWeight: FontWeight.bold,
                ),
              ),
              Text(
                'Camera feed would appear here',
                style: TextStyle(
                  color: Colors.white70,
                  fontSize: 16,
                ),
              ),
            ],
          ),
        ),
      );
    } else {
      return Container(
        color: Colors.black,
        child: Center(
          child: Column(
            mainAxisAlignment: MainAxisAlignment.center,
            children: [
              CircleAvatar(
                radius: 60,
                backgroundImage: NetworkImage(
                  widget.streamerUsername != null
                      ? 'https://example.com/${widget.streamerUsername}.jpg'
                      : 'https://example.com/current_user_avatar.jpg',
                ),
              ),
              SizedBox(height: 20),
              Text(
                widget.streamerUsername ?? 'You',
                style: TextStyle(
                  color: Colors.white,
                  fontSize: 24,
                  fontWeight: FontWeight.bold,
                ),
              ),
              SizedBox(height: 10),
              Container(
                padding: EdgeInsets.symmetric(horizontal: 12, vertical: 6),
                decoration: BoxDecoration(
                  color: Colors.red,
                  borderRadius: BorderRadius.circular(4),
                ),
                child: Text(
                  'LIVE ENDED',
                  style: TextStyle(
                    color: Colors.white,
                    fontSize: 12,
                    fontWeight: FontWeight.bold,
                  ),
                ),
              ),
            ],
          ),
        ),
      );
    }
  }

  Widget _buildTopOverlay() {
    return Container(
      padding: EdgeInsets.all(16),
      child: Column(
        children: [
          Row(
            children: [
              // Live indicator
              AnimatedBuilder(
                animation: _liveIndicatorController,
                builder: (context, child) {
                  return Container(
                    padding: EdgeInsets.symmetric(horizontal: 8, vertical: 4),
                    decoration: BoxDecoration(
                      color: Colors.red.withOpacity(
                        isStreaming
                            ? 0.8 + 0.2 * _liveIndicatorController.value
                            : 0.5,
                      ),
                      borderRadius: BorderRadius.circular(4),
                    ),
                    child: Text(
                      'LIVE',
                      style: TextStyle(
                        color: Colors.white,
                        fontSize: 12,
                        fontWeight: FontWeight.bold,
                      ),
                    ),
                  );
                },
              ),
              SizedBox(width: 8),
              // Viewer count
              Container(
                padding: EdgeInsets.symmetric(horizontal: 8, vertical: 4),
                decoration: BoxDecoration(
                  color: Colors.black45,
                  borderRadius: BorderRadius.circular(12),
                ),
                child: Row(
                  mainAxisSize: MainAxisSize.min,
                  children: [
                    Icon(Icons.visibility, color: Colors.white, size: 16),
                    SizedBox(width: 4),
                    Text(
                      _formatCount(viewerCount),
                      style: TextStyle(color: Colors.white, fontSize: 12),
                    ),
                  ],
                ),
              ),
              Spacer(),
              // Close button
              GestureDetector(
                onTap: () => Navigator.pop(context),
                child: Container(
                  width: 32,
                  height: 32,
                  decoration: BoxDecoration(
                    color: Colors.black45,
                    shape: BoxShape.circle,
                  ),
                  child: Icon(Icons.close, color: Colors.white, size: 20),
                ),
              ),
            ],
          ),
          if (!isStreaming) ...[
            SizedBox(height: 16),
            // Streamer info for ended stream
            Row(
              children: [
                CircleAvatar(
                  radius: 20,
                  backgroundImage: NetworkImage(
                    'https://example.com/${widget.streamerUsername ?? 'current_user'}.jpg',
                  ),
                ),
                SizedBox(width: 12),
                Expanded(
                  child: Column(
                    crossAxisAlignment: CrossAxisAlignment.start,
                    children: [
                      Text(
                        widget.streamerUsername ?? 'You',
                        style: TextStyle(
                          color: Colors.white,
                          fontWeight: FontWeight.bold,
                        ),
                      ),
                      Text(
                        'Stream ended • ${_formatCount(viewerCount)} viewers',
                        style: TextStyle(
                          color: Colors.white70,
                          fontSize: 12,
                        ),
                      ),
                    ],
                  ),
                ),
                if (widget.streamerUsername != null)
                  OutlinedButton(
                    onPressed: _followStreamer,
                    child: Text('Follow', style: TextStyle(color: Colors.white)),
                    style: OutlinedButton.styleFrom(
                      side: BorderSide(color: Colors.white),
                    ),
                  ),
              ],
            ),
          ],
        ],
      ),
    );
  }

  Widget _buildChatOverlay() {
    return Positioned(
      top: MediaQuery.of(context).size.height * 0.3,
      bottom: 120,
      left: 16,
      right: MediaQuery.of(context).size.width * 0.3,
      child: Container(
        decoration: BoxDecoration(
          color: Colors.black26,
          borderRadius: BorderRadius.circular(8),
        ),
        child: Column(
          children: [
            // Chat header
            Container(
              padding: EdgeInsets.all(12),
              decoration: BoxDecoration(
                color: Colors.black54,
                borderRadius: BorderRadius.vertical(top: Radius.circular(8)),
              ),
              child: Row(
                children: [
                  Icon(Icons.chat, color: Colors.white, size: 16),
                  SizedBox(width: 8),
                  Text(
                    'Live Chat',
                    style: TextStyle(
                      color: Colors.white,
                      fontWeight: FontWeight.bold,
                    ),
                  ),
                  Spacer(),
                  GestureDetector(
                    onTap: () => setState(() => showChat = false),
                    child: Icon(Icons.close, color: Colors.white, size: 16),
                  ),
                ],
              ),
            ),
            // Messages
            Expanded(
              child: ListView.builder(
                controller: _chatController,
                padding: EdgeInsets.all(8),
                itemCount: messages.length,
                itemBuilder: (context, index) {
                  return _buildChatMessage(messages[index]);
                },
              ),
            ),
          ],
        ),
      ),
    );
  }

  Widget _buildChatMessage(LiveMessage message) {
    return Container(
      margin: EdgeInsets.only(bottom: 4),
      padding: EdgeInsets.symmetric(horizontal: 8, vertical: 4),
      decoration: BoxDecoration(
        color: message.isHighlighted ? Colors.yellow.withOpacity(0.2) : null,
        borderRadius: BorderRadius.circular(4),
      ),
      child: RichText(
        text: TextSpan(
          style: TextStyle(color: Colors.white, fontSize: 12),
          children: [
            TextSpan(
              text: '${message.username}: ',
              style: TextStyle(
                fontWeight: FontWeight.bold,
                color: message.isHighlighted ? Colors.yellow : Colors.white,
              ),
            ),
            TextSpan(text: message.message),
          ],
        ),
      ),
    );
  }

  Widget _buildBottomControls() {
    return Container(
      padding: EdgeInsets.all(16),
      child: Row(
        children: [
          // Chat input (only if streaming)
          if (isStreaming) ...[
            Expanded(
              child: Container(
                height: 40,
                padding: EdgeInsets.symmetric(horizontal: 12),
                decoration: BoxDecoration(
                  color: Colors.black45,
                  borderRadius: BorderRadius.circular(20),
                  border: Border.all(color: Colors.white30),
                ),
                child: TextField(
                  controller: _messageController,
                  style: TextStyle(color: Colors.white),
                  decoration: InputDecoration(
                    hintText: 'Say something...',
                    hintStyle: TextStyle(color: Colors.white70),
                    border: InputBorder.none,
                  ),
                  onSubmitted: _sendMessage,
                ),
              ),
            ),
            SizedBox(width: 12),
          ],
          // Action buttons
          _buildActionButton(
            icon: showChat ? Icons.chat : Icons.chat_bubble_outline,
            onTap: () => setState(() => showChat = !showChat),
          ),
          SizedBox(width: 12),
          _buildActionButton(
            icon: isLiked ? Icons.favorite : Icons.favorite_border,
            color: isLiked ? Colors.red : Colors.white,
            onTap: _toggleLike,
            count: likeCount,
          ),
          SizedBox(width: 12),
          _buildActionButton(
            icon: Icons.share,
            onTap: _shareStream,
          ),
          if (!isStreaming && widget.streamerUsername == null) ...[
            SizedBox(width: 12),
            ElevatedButton.icon(
              onPressed: _startLiveStream,
              icon: Icon(Icons.videocam),
              label: Text('Go Live'),
              style: ElevatedButton.styleFrom(
                backgroundColor: Colors.red,
                foregroundColor: Colors.white,
              ),
            ),
          ],
          if (isStreaming && widget.streamerUsername == null) ...[
            SizedBox(width: 12),
            ElevatedButton.icon(
              onPressed: _endLiveStream,
              icon: Icon(Icons.stop),
              label: Text('End'),
              style: ElevatedButton.styleFrom(
                backgroundColor: Colors.red,
                foregroundColor: Colors.white,
              ),
            ),
          ],
        ],
      ),
    );
  }

  Widget _buildActionButton({
    required IconData icon,
    required VoidCallback onTap,
    Color color = Colors.white,
    int? count,
  }) {
    return GestureDetector(
      onTap: onTap,
      child: Column(
        mainAxisSize: MainAxisSize.min,
        children: [
          Container(
            width: 48,
            height: 48,
            decoration: BoxDecoration(
              color: Colors.black45,
              shape: BoxShape.circle,
            ),
            child: Icon(icon, color: color),
          ),
          if (count != null) ...[
            SizedBox(height: 4),
            Text(
              _formatCount(count),
              style: TextStyle(
                color: Colors.white,
                fontSize: 12,
                fontWeight: FontWeight.bold,
              ),
            ),
          ],
        ],
      ),
    );
  }

  void _sendMessage(String text) {
    if (text.trim().isNotEmpty) {
      setState(() {
        messages.add(LiveMessage(
          id: DateTime.now().millisecondsSinceEpoch.toString(),
          username: 'You',
          message: text,
          timestamp: DateTime.now(),
          isHighlighted: false,
        ));
      });
      _messageController.clear();
      _scrollToBottom();
    }
  }

  void _toggleLike() {
    setState(() {
      isLiked = !isLiked;
      likeCount += isLiked ? 1 : -1;
    });

    if (isLiked) {
      // Show floating heart animation
      _showFloatingHeart();
    }
  }

  void _showFloatingHeart() {
    // This would show a floating heart animation
    ScaffoldMessenger.of(context).showSnackBar(
      SnackBar(
        content: Text('❤️ Liked!'),
        duration: Duration(milliseconds: 800),
        backgroundColor: Colors.transparent,
        elevation: 0,
      ),
    );
  }

  void _shareStream() {
    showModalBottomSheet(
      context: context,
      backgroundColor: Colors.white,
      shape: RoundedRectangleBorder(
        borderRadius: BorderRadius.vertical(top: Radius.circular(16)),
      ),
      builder: (context) {
        return Container(
          padding: EdgeInsets.all(16),
          child: Column(
            mainAxisSize: MainAxisSize.min,
            children: [
              Text(
                'Share Live Stream',
                style: TextStyle(
                  fontSize: 18,
                  fontWeight: FontWeight.bold,
                ),
              ),
              SizedBox(height: 20),
              ListTile(
                leading: Icon(Icons.copy),
                title: Text('Copy Link'),
                onTap: () {
                  Navigator.pop(context);
                  ScaffoldMessenger.of(context).showSnackBar(
                    SnackBar(content: Text('Live stream link copied')),
                  );
                },
              ),
              ListTile(
                leading: Icon(Icons.share),
                title: Text('Share to Story'),
                onTap: () {
                  Navigator.pop(context);
                  ScaffoldMessenger.of(context).showSnackBar(
                    SnackBar(content: Text('Shared to your story')),
                  );
                },
              ),
              ListTile(
                leading: Icon(Icons.send),
                title: Text('Send to Friends'),
                onTap: () {
                  Navigator.pop(context);
                  // Show friends list
                },
              ),
            ],
          ),
        );
      },
    );
  }

  void _startLiveStream() {
    setState(() {
      isStreaming = true;
      viewerCount = 1;
    });
    _liveIndicatorController.repeat();
    _simulateMessages();

    ScaffoldMessenger.of(context).showSnackBar(
      SnackBar(
        content: Text('Live stream started! 🔴'),
        backgroundColor: Colors.green,
      ),
    );
  }

  void _endLiveStream() {
    showDialog(
      context: context,
      builder: (context) => AlertDialog(
        title: Text('End Live Stream?'),
        content: Text('Are you sure you want to end your live stream?'),
        actions: [
          TextButton(
            onPressed: () => Navigator.pop(context),
            child: Text('Cancel'),
          ),
          TextButton(
            onPressed: () {
              Navigator.pop(context);
              setState(() {
                isStreaming = false;
              });
              _liveIndicatorController.stop();

              ScaffoldMessenger.of(context).showSnackBar(
                SnackBar(
                  content: Text('Live stream ended'),
                  backgroundColor: Colors.grey,
                ),
              );
            },
            child: Text('End Stream', style: TextStyle(color: Colors.red)),
          ),
        ],
      ),
    );
  }

  void _followStreamer() {
    ScaffoldMessenger.of(context).showSnackBar(
      SnackBar(
        content: Text('Following ${widget.streamerUsername}'),
        backgroundColor: Colors.blue,
      ),
    );
  }

  void _scrollToBottom() {
    if (_chatController.hasClients) {
      _chatController.animateTo(
        _chatController.position.maxScrollExtent,
        duration: Duration(milliseconds: 300),
        curve: Curves.easeOut,
      );
    }
  }

  String _formatCount(int count) {
    if (count < 1000) return count.toString();
    if (count < 1000000) return '${(count / 1000).toStringAsFixed(1)}K';
    return '${(count / 1000000).toStringAsFixed(1)}M';
  }

  @override
  void dispose() {
    _messageController.dispose();
    _chatController.dispose();
    _liveIndicatorController.dispose();
    super.dispose();
  }
}

class LiveMessage {
  final String id;
  final String username;
  final String message;
  final DateTime timestamp;
  final bool isHighlighted;

  LiveMessage({
    required this.id,
    required this.username,
    required this.message,
    required this.timestamp,
    required this.isHighlighted,
  });
}

File: snippets/screens/swiftui/live_stream_screen/2/live_stream_screen.swift
import SwiftUI

struct LiveStreamScreen: View {
    let streamerUsername: String?
    let isStreaming: Bool

    @StateObject private var viewModel = LiveStreamViewModel()
    @Environment(\.dismiss) private var dismiss
    @State private var messageText = ""
    @FocusState private var isMessageFieldFocused: Bool

    var body: some View {
        ZStack {
            Color.black.ignoresSafeArea()

            // Video stream background
            StreamBackgroundView(
                isStreaming: viewModel.isStreaming,
                streamerUsername: streamerUsername ?? "current_user"
            )

            VStack(spacing: 0) {
                // Top overlay
                StreamHeaderView(
                    isStreaming: viewModel.isStreaming,
                    viewerCount: viewModel.viewerCount,
                    streamerUsername: streamerUsername,
                    onClose: { dismiss() }
                )
                .padding(.horizontal, 16)
                .padding(.top, 8)

                Spacer()

                // Chat overlay
                if viewModel.showChat {
                    ChatOverlayView(
                        messages: viewModel.messages,
                        onCloseChat: { viewModel.showChat = false }
                    )
                    .frame(maxHeight: 300)
                    .padding(.horizontal, 16)
                    .padding(.bottom, 120)
                }

                Spacer()

                // Bottom controls
                StreamBottomControls(
                    isStreaming: viewModel.isStreaming,
                    showChat: viewModel.showChat,
                    likeCount: viewModel.likeCount,
                    isLiked: viewModel.isLiked,
                    messageText: $messageText,
                    isMessageFieldFocused: $isMessageFieldFocused,
                    streamerUsername: streamerUsername,
                    onToggleChat: { viewModel.showChat.toggle() },
                    onLike: viewModel.toggleLike,
                    onShare: viewModel.shareStream,
                    onSendMessage: { text in
                        viewModel.sendMessage(text)
                        messageText = ""
                    },
                    onStartStream: viewModel.startLiveStream,
                    onEndStream: viewModel.showEndStreamAlert
                )
                .padding(.horizontal, 16)
                .padding(.bottom, 16)
            }

            // Floating hearts animation
            if viewModel.showFloatingHeart {
                FloatingHeartView()
                    .onAppear {
                        DispatchQueue.main.asyncAfter(deadline: .now() + 2) {
                            viewModel.showFloatingHeart = false
                        }
                    }
            }
        }
        .navigationBarHidden(true)
        .onAppear {
            viewModel.initialize(
                streamerUsername: streamerUsername,
                isStreaming: isStreaming
            )
        }
        .alert("End Live Stream?", isPresented: $viewModel.showEndAlert) {
            Button("Cancel", role: .cancel) { }
            Button("End Stream", role: .destructive) {
                viewModel.endLiveStream()
                dismiss()
            }
        } message: {
            Text("Are you sure you want to end your live stream?")
        }
        .sheet(isPresented: $viewModel.showShareSheet) {
            ShareStreamSheet()
        }
    }
}

struct StreamBackgroundView: View {
    let isStreaming: Bool
    let streamerUsername: String

    var body: some View {
        if isStreaming {
            // Live stream background with gradient
            Rectangle()
                .fill(
                    LinearGradient(
                        colors: [
                            Color.purple.opacity(0.3),
                            Color.black,
                            Color.blue.opacity(0.3)
                        ],
                        startPoint: .top,
                        endPoint: .bottom
                    )
                )
                .overlay(
                    VStack {
                        Image(systemName: "video.fill")
                            .font(.system(size: 120))
                            .foregroundColor(.white.opacity(0.5))

                        Text("Live Stream")
                            .font(.title)
                            .fontWeight(.bold)
                            .foregroundColor(.white)

                        Text("Camera feed would appear here")
                            .font(.body)
                            .foregroundColor(.white.opacity(0.7))
                    }
                )
        } else {
            // Stream ended background
            VStack(spacing: 20) {
                AsyncImage(url: URL(string: "https://example.com/\(streamerUsername).jpg")) { image in
                    image
                        .resizable()
                        .aspectRatio(contentMode: .fill)
                } placeholder: {
                    Circle()
                        .fill(Color.gray.opacity(0.3))
                }
                .frame(width: 120, height: 120)
                .clipShape(Circle())

                Text(streamerUsername)
                    .font(.title)
                    .fontWeight(.bold)
                    .foregroundColor(.white)

                HStack {
                    Text("LIVE ENDED")
                        .font(.caption)
                        .fontWeight(.bold)
                        .foregroundColor(.white)
                        .padding(.horizontal, 12)
                        .padding(.vertical, 6)
                        .background(Color.red.opacity(0.8))
                        .cornerRadius(4)
                }
            }
        }
    }
}

struct StreamHeaderView: View {
    let isStreaming: Bool
    let viewerCount: Int
    let streamerUsername: String?
    let onClose: () -> Void

    var body: some View {
        HStack {
            // Live indicator
            HStack(spacing: 8) {
                LiveIndicatorView(isStreaming: isStreaming)

                // Viewer count
                HStack(spacing: 4) {
                    Image(systemName: "eye")
                        .font(.caption)
                    Text(formatCount(viewerCount))
                        .font(.caption)
                }
                .foregroundColor(.white)
                .padding(.horizontal, 8)
                .padding(.vertical, 4)
                .background(Color.black.opacity(0.5))
                .cornerRadius(12)
            }

            Spacer()

            // Close button
            Button(action: onClose) {
                Image(systemName: "xmark")
                    .font(.title2)
                    .foregroundColor(.white)
                    .frame(width: 32, height: 32)
                    .background(Color.black.opacity(0.5))
                    .clipShape(Circle())
            }
        }
    }

    private func formatCount(_ count: Int) -> String {
        if count < 1000 { return "\(count)" }
        if count < 1000000 { return String(format: "%.1fK", Double(count) / 1000) }
        return String(format: "%.1fM", Double(count) / 1000000)
    }
}

struct LiveIndicatorView: View {
    let isStreaming: Bool
    @State private var pulseScale: CGFloat = 1.0

    var body: some View {
        Text("LIVE")
            .font(.caption)
            .fontWeight(.bold)
            .foregroundColor(.white)
            .padding(.horizontal, 8)
            .padding(.vertical, 4)
            .background(Color.red.opacity(isStreaming ? 0.9 : 0.5))
            .cornerRadius(4)
            .scaleEffect(pulseScale)
            .animation(
                isStreaming ?
                Animation.easeInOut(duration: 1.0).repeatForever(autoreverses: true) :
                .none,
                value: pulseScale
            )
            .onAppear {
                if isStreaming {
                    pulseScale = 1.1
                }
            }
    }
}

struct ChatOverlayView: View {
    let messages: [LiveMessage]
    let onCloseChat: () -> Void

    var body: some View {
        VStack(spacing: 0) {
            // Chat header
            HStack {
                HStack(spacing: 8) {
                    Image(systemName: "bubble.left.and.bubble.right.fill")
                        .font(.caption)
                    Text("Live Chat")
                        .font(.subheadline)
                        .fontWeight(.semibold)
                }
                .foregroundColor(.white)

                Spacer()

                Button(action: onCloseChat) {
                    Image(systemName: "xmark")
                        .font(.caption)
                        .foregroundColor(.white)
                }
            }
            .padding(.all, 12)
            .background(Color.black.opacity(0.6))

            // Messages
            ScrollView {
                LazyVStack(alignment: .leading, spacing: 4) {
                    ForEach(messages) { message in
                        ChatMessageView(message: message)
                    }
                }
                .padding(.horizontal, 8)
                .padding(.vertical, 8)
            }
            .background(Color.black.opacity(0.3))
        }
        .cornerRadius(8)
    }
}

struct ChatMessageView: View {
    let message: LiveMessage

    var body: some View {
        HStack(alignment: .top) {
            Text(message.username + ":")
                .font(.caption)
                .fontWeight(.bold)
                .foregroundColor(message.isHighlighted ? .yellow : .white)
            + Text(" " + message.message)
                .font(.caption)
                .foregroundColor(.white)

            Spacer(minLength: 0)
        }
        .padding(.horizontal, 8)
        .padding(.vertical, 4)
        .background(
            message.isHighlighted ?
            Color.yellow.opacity(0.2) :
            Color.clear
        )
        .cornerRadius(4)
    }
}

struct StreamBottomControls: View {
    let isStreaming: Bool
    let showChat: Bool
    let likeCount: Int
    let isLiked: Bool
    @Binding var messageText: String
    var isMessageFieldFocused: FocusState<Bool>.Binding
    let streamerUsername: String?

    let onToggleChat: () -> Void
    let onLike: () -> Void
    let onShare: () -> Void
    let onSendMessage: (String) -> Void
    let onStartStream: () -> Void
    let onEndStream: () -> Void

    var body: some View {
        HStack(spacing: 12) {
            // Message input (only when streaming)
            if isStreaming {
                HStack(spacing: 8) {
                    TextField("Say something...", text: $messageText)
                        .focused(isMessageFieldFocused)
                        .textFieldStyle(.plain)
                        .foregroundColor(.white)
                        .onSubmit {
                            if !messageText.trimmingCharacters(in: .whitespacesAndNewlines).isEmpty {
                                onSendMessage(messageText)
                            }
                        }
                }
                .padding(.horizontal, 12)
                .padding(.vertical, 8)
                .background(Color.black.opacity(0.5))
                .cornerRadius(20)
                .overlay(
                    RoundedRectangle(cornerRadius: 20)
                        .stroke(Color.white.opacity(0.3), lineWidth: 1)
                )
            }

            // Action buttons
            StreamActionButton(
                icon: showChat ? "bubble.left.and.bubble.right.fill" : "bubble.left.and.bubble.right",
                onTap: onToggleChat
            )

            StreamActionButton(
                icon: isLiked ? "heart.fill" : "heart",
                color: isLiked ? .red : .white,
                count: likeCount,
                onTap: onLike
            )

            StreamActionButton(
                icon: "square.and.arrow.up",
                onTap: onShare
            )

            // Stream control buttons
            if !isStreaming && streamerUsername == nil {
                Button("Go Live") {
                    onStartStream()
                }
                .font(.subheadline)
                .fontWeight(.semibold)
                .foregroundColor(.white)
                .padding(.horizontal, 16)
                .padding(.vertical, 8)
                .background(Color.red)
                .cornerRadius(8)
            }

            if isStreaming && streamerUsername == nil {
                Button("End") {
                    onEndStream()
                }
                .font(.subheadline)
                .fontWeight(.semibold)
                .foregroundColor(.white)
                .padding(.horizontal, 16)
                .padding(.vertical, 8)
                .background(Color.red)
                .cornerRadius(8)
            }
        }
    }
}

struct StreamActionButton: View {
    let icon: String
    let color: Color
    let count: Int?
    let onTap: () -> Void

    init(icon: String, color: Color = .white, count: Int? = nil, onTap: @escaping () -> Void) {
        self.icon = icon
        self.color = color
        self.count = count
        self.onTap = onTap
    }

    var body: some View {
        Button(action: onTap) {
            VStack(spacing: 4) {
                Circle()
                    .fill(Color.black.opacity(0.5))
                    .frame(width: 48, height: 48)
                    .overlay(
                        Image(systemName: icon)
                            .font(.title2)
                            .foregroundColor(color)
                    )

                if let count = count {
                    Text(formatCount(count))
                        .font(.caption)
                        .fontWeight(.bold)
                        .foregroundColor(.white)
                }
            }
        }
    }

    private func formatCount(_ count: Int) -> String {
        if count < 1000 { return "\(count)" }
        if count < 1000000 { return String(format: "%.1fK", Double(count) / 1000) }
        return String(format: "%.1fM", Double(count) / 1000000)
    }
}

struct FloatingHeartView: View {
    @State private var offset = CGSize.zero
    @State private var opacity = 1.0
    @State private var scale = 0.1

    var body: some View {
        VStack {
            Spacer()
            HStack {
                Spacer()
                Image(systemName: "heart.fill")
                    .font(.title)
                    .foregroundColor(.red)
                    .scaleEffect(scale)
                    .opacity(opacity)
                    .offset(offset)
                    .onAppear {
                        withAnimation(.easeOut(duration: 2.0)) {
                            offset = CGSize(width: 0, height: -300)
                            opacity = 0.0
                            scale = 1.5
                        }
                    }
                Spacer()
            }
        }
    }
}

struct ShareStreamSheet: View {
    @Environment(\.dismiss) private var dismiss

    var body: some View {
        NavigationView {
            VStack(spacing: 20) {
                Text("Share Live Stream")
                    .font(.title2)
                    .fontWeight(.bold)
                    .padding(.top, 20)

                VStack(spacing: 16) {
                    ShareOption(
                        icon: "doc.on.doc",
                        title: "Copy Link",
                        action: {
                            dismiss()
                        }
                    )

                    ShareOption(
                        icon: "plus.rectangle.on.rectangle",
                        title: "Share to Story",
                        action: {
                            dismiss()
                        }
                    )

                    ShareOption(
                        icon: "paperplane",
                        title: "Send to Friends",
                        action: {
                            dismiss()
                        }
                    )
                }
                .padding(.horizontal, 20)

                Spacer()
            }
            .navigationTitle("Share")
            .navigationBarTitleDisplayMode(.inline)
            .toolbar {
                ToolbarItem(placement: .navigationBarTrailing) {
                    Button("Cancel") {
                        dismiss()
                    }
                }
            }
        }
    }
}

struct ShareOption: View {
    let icon: String
    let title: String
    let action: () -> Void

    var body: some View {
        Button(action: action) {
            HStack {
                Image(systemName: icon)
                    .font(.title2)
                    .frame(width: 24)
                Text(title)
                    .font(.body)
                Spacer()
                Image(systemName: "chevron.right")
                    .font(.caption)
                    .foregroundColor(.secondary)
            }
            .foregroundColor(.primary)
            .padding(.vertical, 12)
        }
    }
}

@MainActor
class LiveStreamViewModel: ObservableObject {
    @Published var isStreaming = false
    @Published var viewerCount = 1247
    @Published var likeCount = 3456
    @Published var isLiked = false
    @Published var showChat = true
    @Published var messages: [LiveMessage] = []
    @Published var showFloatingHeart = false
    @Published var showEndAlert = false
    @Published var showShareSheet = false

    private var messageSimulationTimer: Timer?

    func initialize(streamerUsername: String?, isStreaming: Bool) {
        self.isStreaming = isStreaming

        // Sample messages
        messages = [
            LiveMessage(
                id: "1",
                username: "viewer1",
                message: "Great stream! 🔥",
                timestamp: Date().addingTimeInterval(-120),
                isHighlighted: false
            ),
            LiveMessage(
                id: "2",
                username: "viewer2",
                message: "Hello everyone! 👋",
                timestamp: Date().addingTimeInterval(-60),
                isHighlighted: false
            ),
            LiveMessage(
                id: "3",
                username: "moderator",
                message: "Welcome to the stream!",
                timestamp: Date().addingTimeInterval(-30),
                isHighlighted: true
            )
        ]

        if isStreaming {
            startMessageSimulation()
        }
    }

    func toggleLike() {
        isLiked.toggle()
        likeCount += isLiked ? 1 : -1

        if isLiked {
            showFloatingHeart = true
        }
    }

    func sendMessage(_ text: String) {
        let newMessage = LiveMessage(
            id: UUID().uuidString,
            username: "You",
            message: text,
            timestamp: Date(),
            isHighlighted: false
        )

        messages.append(newMessage)
    }

    func shareStream() {
        showShareSheet = true
    }

    func startLiveStream() {
        isStreaming = true
        viewerCount = 1
        startMessageSimulation()
    }

    func showEndStreamAlert() {
        showEndAlert = true
    }

    func endLiveStream() {
        isStreaming = false
        stopMessageSimulation()
    }

    private func startMessageSimulation() {
        messageSimulationTimer = Timer.scheduledTimer(withTimeInterval: 5.0, repeats: true) { _ in
            if self.isStreaming {
                let newMessage = LiveMessage(
                    id: UUID().uuidString,
                    username: "random_viewer",
                    message: "This is amazing! 😍",
                    timestamp: Date(),
                    isHighlighted: false
                )
                self.messages.append(newMessage)
                self.viewerCount += Int.random(in: 0...5)
            }
        }
    }

    private func stopMessageSimulation() {
        messageSimulationTimer?.invalidate()
        messageSimulationTimer = nil
    }

    deinit {
        stopMessageSimulation()
    }
}

struct LiveMessage: Identifiable {
    let id: String
    let username: String
    let message: String
    let timestamp: Date
    let isHighlighted: Bool
}

#Preview {
    LiveStreamScreen(streamerUsername: nil, isStreaming: true)
}